ciongke/cmd/test-pr/main.go

```diff
@@ -0,0 +1,395 @@
+		return false, err
+	}
+	mergeCommand.Dir = clonePath
+	if err := cmdlog.RunWithLogs(mergeCommand); err != nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+func test(tc testClient, j jobDescription, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	jobPath := fmt.Sprintf("pull/%d/%s", *pr, j.Name)
+	buildNum, err := gcs.IncrementAndGetBuildNumber(*resultsBucket, jobPath)
+	if err != nil {
+		log.Printf("Could not get build number: %s", err)
+		return
+	}
+	buildPath := fmt.Sprintf("%s/%d", jobPath, buildNum)
+	if err = gcs.UploadStarted(*resultsBucket, buildPath, *head); err != nil {
+		log.Printf("Could not upload started.json: %s", err)
+		return
+	}
+	err = tc.GitHubClient.CreateStatus(*repoOwner, *repoName, *head, github.Status{
+		State:       github.Pending,
+		Context:     "ciongke/" + j.Name,
+		Description: j.Description + " started",
+	})
+	if err != nil {
+		log.Printf("Error creating status: %s", err)
+		return
+	}
+
+	buildLog, exitCode, err := runTest(tc, j)
+	if err != nil {
+		// TODO: Handle this error. For now it will call the test a success.
+		log.Printf("Error running test: %s", err)
+	} else if err = gcs.UploadToGCS(buildLog, *resultsBucket, buildPath+"/build-log.txt"); err != nil {
+		log.Printf("Could not upload build log: %s", err)
+	}
+
+	testResult := gcs.Unknown
+	status := github.Error
+	if exitCode == 0 {
+		testResult = gcs.Success
+		status = github.Success
+	} else {
+		testResult = gcs.Failure
+		status = github.Failure
+	}
+
+	if err = gcs.UploadFinished(*resultsBucket, buildPath, testResult); err != nil {
+		log.Printf("Could not upload finished.json: %s", err)
+	}
+
+	testURL := fmt.Sprintf("https://storage.googleapis.com/%s/pull/%d/%s/%d/build-log.txt", *resultsBucket, *pr, j.Name, buildNum)
+	err = tc.GitHubClient.CreateStatus(*repoOwner, *repoName, *head, github.Status{
+		State:       status,
+		Context:     "ciongke/" + j.Name,
+		Description: fmt.Sprintf("%s: %s", j.Description, status),
+		TargetURL:   testURL,
+	})
+	if err != nil {
+		log.Printf("Could not set PR status: %s", err)
+	}
+}
+
+// runTest starts a job based on the job description. When the job completes,
+// it returns its log and exit code.
+func runTest(tc testClient, j jobDescription) (*bytes.Buffer, int, error) {
+	ji := tc.KubeBatchClient.Jobs(*namespace)
+	pi := tc.KubeClient.Pods(*namespace)
+	w, err := ji.Watch(api.ListOptions{
+		Watch: true,
+		LabelSelector: labels.SelectorFromSet(labels.Set{
+			"repo": *repoName,
+			"pr":   strconv.Itoa(*pr),
+			"test": j.Name,
+		}),
+	})
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not watch job: %s", err)
+	}
+	result := make(chan error)
+	go watchTest(w, j.Name, result)
+
+	priv := true
+	job := &batch.Job{
+		ObjectMeta: api.ObjectMeta{
+			Name:      fmt.Sprintf("%s-pr-%d-%s-%s", *repoName, *pr, (*head)[:8], j.Name),
+			Namespace: *namespace,
+			Labels: map[string]string{
+				"repo": *repoName,
+				"pr":   strconv.Itoa(*pr),
```

Comment:
fyi: `pr` makes me think this should be something defined at the top of the function rather than a flag :(